pkg/reconciler/v1alpha1/taskrun/taskrun.go

```diff
@@ -504,24 +518,48 @@ func (c *Reconciler) checkTimeout(tr *v1alpha1.TaskRun, ts *v1alpha1.TaskSpec, d
 		timeout := tr.Spec.Timeout.Duration
 		runtime := time.Since(tr.Status.StartTime.Time)
 		if runtime > timeout {
-			c.Logger.Infof("TaskRun %q is timeout (runtime %s over %s), deleting pod", tr.Name, runtime, timeout)
-			if err := dp(tr.Status.PodName, &metav1.DeleteOptions{}); err != nil && !errors.IsNotFound(err) {
-				c.Logger.Errorf("Failed to terminate pod: %v", err)
-				return true, err
-			}
 
+			status := tr.Status.DeepCopy()
 			timeoutMsg := fmt.Sprintf("TaskRun %q failed to finish within %q", tr.Name, timeout.String())
-			tr.Status.SetCondition(&duckv1alpha1.Condition{
+			status.SetCondition(&duckv1alpha1.Condition{
 				Type:    duckv1alpha1.ConditionSucceeded,
 				Status:  corev1.ConditionFalse,
 				Reason:  reasonTimedOut,
 				Message: timeoutMsg,
 			})
 			// update tr completed time
-			tr.Status.CompletionTime = &metav1.Time{Time: time.Now()}
+			status.CompletionTime = &metav1.Time{Time: time.Now()}
+
+			if retry(tr, status) {
+				return false, nil
+			}
+
+			c.Logger.Infof("TaskRun %q is timeout (runtime %s over %s), deleting pod", tr.Name, runtime, timeout)
+			if err := dp(tr.Status.PodName, &metav1.DeleteOptions{}); err != nil && !errors.IsNotFound(err) {
+				c.Logger.Errorf("Failed to terminate pod: %v", err)
+				return true, err
+			}
 
 			return true, nil
 		}
 	}
 	return false, nil
 }
+
+func retry(tr *v1alpha1.TaskRun, status *v1alpha1.TaskRunStatus) bool {
+
+	if len(tr.Status.Retries) < tr.Spec.Retries {
+		tr.Status.StartTime = nil
```

Comment:
In `taskrunStatus` the only field to retain is `Retries` & `Condition` but rest of the fields you can wipe out everything to nil right? 
Why only set `StartTime` & `CompletionTime` to nil ?

Reply:
what about `podName`?

Reply:
Interesting point. Do we want new retry to have same podname or diff? 
Benefits I could see for new podName would be 
- If pod is not deleted properly then creation of new pod with same name can fail.
- If user wants to fetch logs from failed pods they can do so.

This could be an enhancement to retry feature. Something to think about.

Reply:
Under some circumstances reusing the pod could be nice, but having in mind k8s ephemeral approach would be nice to always ditch it. WYDY @shashwathi ?

Reply:
Can you explain what do you mean by "Under some circumstances reusing the pod could be nice"? I am all for new pod name on each retry.

Reply:
I think a new pod is always better. Reusing the pod could be good, but tough to find out.
I will do it with a new pod when retrying